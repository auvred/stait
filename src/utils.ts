export type ContextDecl = Record<string, unknown>

export type ActionsDecl<_Typestait extends { type: string }> = Record<
  string,
  // `any` looks cleaner in user-facing errors
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (payload?: any) => _Typestait['type']
>

export type Flatten<T> = [T] extends [infer TT]
  ? {
      [K in keyof TT]: TT[K]
    }
  : never
