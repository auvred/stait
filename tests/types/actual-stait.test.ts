import { runTypesTests } from '../utils'

import type { ActualStaitFromTypestait, Typestait } from '../../src'

runTypesTests(import.meta.url)

type ExpectType_actual_stait = ActualStaitFromTypestait<
  Typestait<
    | {
        type: 'stait1'
        context: {
          data1: string
          data2: 123
        }
        actions: {
          action1: (payload: number) => 'stait1' | 'stait2'
        }
      }
    | {
        type: 'stait2'
        context: {
          data3: string
          data4: 123
        }
        service: 'stait1' | 'stait2'
      }
  >,
  'stait1'
>

type ExpectType_multiple_types = ActualStaitFromTypestait<
  Typestait<
    | {
        type: 'stait1'
      }
    | {
        type: 'stait2'
      }
    | {
        type: 'stait3'
      }
  >,
  'stait1' | 'stait2'
>
