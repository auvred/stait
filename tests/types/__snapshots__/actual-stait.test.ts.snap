// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`type: actual stait 1`] = `
"type actual_stait =
  {
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
    actions: {
      action1: (opts: {
        ctx: {
          data1: string
          data2: 123
        }
        payload: number
      }) => ActualStaitFromTypestait<
        Typestait<
          | {
              type: 'stait1'
              context: {
                data1: string
                data2: 123
              }
              actions: {
                action1: (
                  payload: number,
                ) =>
                  | 'stait1'
                  | 'stait2'
              }
            }
          | {
              type: 'stait2'
              context: {
                data3: string
                data4: 123
              }
              service:
                | 'stait1'
                | 'stait2'
            }
        >,
        | 'stait1'
        | 'stait2'
      >
    }
  }
"
`;

exports[`type: multiple types 1`] = `
"type multiple_types =

    | {
        type: 'stait1'
      }
    | {
        type: 'stait2'
      }
"
`;
